added working bookstack-traefik installation

xFuture603 opened the pull request:
This PR contains a working state of a bookstack installation, using Traefik as a reverse proxy. We use docker-compose to set up the whole thing.

On top we add renovate to keep track of new software versions to update our compose file accordingly.

tuhlmann commented on README.md:
```diff
@@ -1,19 +1,77 @@
+
+4. Adjust the variables for your configuration. Edit the following values:
+   1. Passwords (declared as "MODIFY_ME")
+   2. E-Mail address (declared as "example.mail@mail.com")
```
Is mail.com a reserved example domain, like example.com, or could this end up in a real inbox?

xFuture603 replied:
I've updated the example mail to something non-existent. Thanks for the hint.

tuhlmann commented on README.md:
```diff
@@ -1,19 +1,77 @@
+   1. Passwords (declared as "MODIFY_ME")
+   2. E-Mail address (declared as "example.mail@mail.com")
+   3. Domain (declared as "example.url.com")
+   4. Encrypted basic auth string for your Traefik dashboard (declared as "admin:$$2y$$10$$tzGrxnBWcBBCKpFWdvzJDOt/Y4ihpuxXuE9r7mP/eo1MRBGcmV9q2")
```
How does the user create their encrypted basic auth string. Should that described here? What's the current value of this string? Wouldn't it be safer to put no value at all in. That way the lazy user cannot just use the password and opens up a security hole, but is forced to create their own password?

xFuture603 replied:
I've added a documentation part on how you can encrypt a password to use it in your basic auth.

tuhlmann commented on traefik-letsencrypt-bookstack-docker-compose.yml:
```diff
@@ -0,0 +1,114 @@
+      # Basic Authentication for Traefik Dashboard
+      # Username: admin (replace with your own username and encrypted password)
+      # Passwords must be encoded using MD5, SHA1, or BCrypt
+      - "traefik.http.middlewares.authtraefik.basicauth.users=admin:$$2y$$10$$tzGrxnBWcBBCKpFWdvzJDOt/Y4ihpuxXuE9r7mP/eo1MRBGcmV9q2"
```
Is there an alternative to putting password in the compose file?

xFuture603 replied:
Sure, you can add a .env file for example or you can vault the hashed password on top with Ansible for example. In my opinion it is okay to leave the hashed password in the compose file because it is already encrypted.

tuhlmann reviewed (changes requested):
Asked a few questions I'd like to have resolved before approval.

tuhlmann commented on traefik-letsencrypt-bookstack-docker-compose.yml:
```diff
@@ -0,0 +1,114 @@
+      # Basic Authentication for Traefik Dashboard
+      # Username: admin (replace with your own username and encrypted password)
+      # Passwords must be encoded using MD5, SHA1, or BCrypt (check README.md on how to hash a password)
+      - "traefik.http.middlewares.authtraefik.basicauth.users=admin:$$2y$$10$$tzGrxnBWcBBCKpFWdvzJDOt/Y4ihpuxXuE9r7mP/eo1MRBGcmV9q2"
```
Sorry to bother you, but it doesn't feel right to me to have passwords- even encrypted ones- in a docker-compose. The reasons might be subtle. I view docker-compose (or any other script) as a rather stable receipt the computer can execute. It shouldn't be touched if a password needs to be updated.
If you could add a `.env.example` file with the key for the password entry in it and a comment but no value, and then instructions how to copy and create that password. I want
- the user having to take action and create their own password, not take a public one because their lazy
- move the password out of the compose and just have compose reference the .env key. Now, I have no idea what you have to do to make compose read passwords from env- I hope that's somewhat straight forward? But you suggested that approach :)

xFuture603 replied:
Sorry for the delay in this PR, due to holidays.

I've added a .env file and moved the basic_auth user and password to it. I've added a unencoded string to the password variable (won't work without encoding to bcrypt for example).

Now you have variables that in the docker-compose file that reference to the .env variables. I hope that everything is fine now. :)

tuhlmann reviewed (changes requested):
Just one issue left I have with this password in docker-compose. Can you address my comment? If you feel strongly that you shouldn't let's chat about it...

tuhlmann reviewed (approved):
Thank you for addressing these issues!!

I have a few comments for future improvements:
- Add a note to the docker-compose comment about which encryption alogorithms can be used- add that BCrypt is strongly recommended
- We normally use upper case (and separated by underscore) environment variable. I'm not sure how this is handled in the Docker/Ansible world. I'm fine with a different standard, but if the same convention applies to Docker/Ansible I suggest with switch out .env variables to UPPER_CASE as well.

xFuture603 commented:
> Thank you for addressing these issues!!
> 
> I have a few comments for future improvements:
> 
> * Add a note to the docker-compose comment about which encryption alogorithms can be used- add that BCrypt is strongly recommended
> * We normally use upper case (and separated by underscore) environment variable. I'm not sure how this is handled in the Docker/Ansible world. I'm fine with a different standard, but if the same convention applies to Docker/Ansible I suggest with switch out .env variables to UPPER_CASE as well.

I've addressed both issues and added them to the PR. I will merge this PR now.
